File: lumina/web/src/api/work-items.ts
// Work-item wire schemas + fetch wrappers.
//
// Carved out of the monolithic `lumina/web/src/api.ts` by T7 of the round-4
// plan (docs/plans/lumina-story-planning-round-4.md). This module owns the
// `WorkItem` / `WorkItemNode` / `WorkItemDetail` shapes and the four
// associated CRUD verbs (`fetchTree`, `fetchDetail`, `createWorkItem`,
// `updateWorkItem`, `updateStatus`).
//
// Nullability convention: every `Option<T>` on `domain::WorkItem` (and every
// other Rust read-aggregate folded into `WorkItemDetail`) uses `.nullable()`
// here. Verified at T7-implementation time by grepping `lumina/src/` for
// `skip_serializing_if` and finding zero hits — serde therefore emits `None`
// as JSON `null` rather than omitting the key. If a future Rust change adds
// `#[serde(skip_serializing_if = "Option::is_none")]` to any field on these
// types, the corresponding zod field must switch to `.nullish()` to tolerate
// the now-absent key.
//
// Cross-task note: this file inline-declares `AcceptanceCriterion`,
// `ResearchNote`, `OpenQuestion`, `QuestionOption`, `Risk`,
// `RejectedAlternative`, `TaskDependency`, `ContextBlock`, `Finding`,
// `WorkItemActivity`, `RepoLink` schemas because `WorkItemDetailWireSchema`
// references them. T9 / T10 / T11 will MOVE the per-family schemas out into
// their owned files (`acceptance-criteria.ts`, `research-notes.ts`,
// `risks.ts`, `rejected-alternatives.ts`, `task-deps.ts`, `findings.ts`,
// `activity.ts`, `open-questions.ts`, `context-blocks.ts`) and re-import them
// here. `RepoLink` lives in its own `repo-links.ts` (carved by T7); we
// re-import it from there.

import * as z from 'zod'

import { API_BASE, handle } from './http'
import {
  type Kind,
  KindSchema,
  type Status,
  StatusSchema,
  type Relevance,
  RelevanceSchema,
  type Effort,
  EffortSchema,
  type Complexity,
  ComplexitySchema,
  type Origin,
  OriginSchema,
  type ClosureGate,
  ClosureGateSchema,
  type Severity,
  SeveritySchema,
  type ActivityType,
  ActivityTypeSchema,
  type ResearchState,
  ResearchStateSchema,
  type QuestionStatus,
  QuestionStatusSchema,
  type Confidence,
  ConfidenceSchema,
  type TaskKind,
  TaskKindSchema,
  type Tier,
  TierSchema,
  type RiskSeverity,
  RiskSeveritySchema,
} from './wire-enums'
import { type RepoLink, RepoLinkSchema } from './repo-links'

// ---------------------------------------------------------------------------
// WorkItem (the row-level read aggregate) + nested-tree variant.
// ---------------------------------------------------------------------------

/** A node in the `work_items` adjacency-list hierarchy. */
export interface WorkItem {
  id: string
  kind: Kind
  parent_id: string | null
  title: string
  body: string | null
  status: Status
  position: number | null
  attributes: Record<string, unknown> | null
  relevance: Relevance | null
  effort: Effort | null
  complexity: Complexity | null
  origin: Origin | null
  closure_gate: ClosureGate | null
  blocked_by_question_id: string | null
  enabling_option_id: string | null
  // Round-4 additions (T7): the round-2/3 dispatch columns.
  task_kind: TaskKind | null
  tier: Tier | null
  created_at: string
  updated_at: string
}

export const WorkItemSchema = z.object({
  id: z.string(),
  kind: KindSchema,
  parent_id: z.string().nullable(),
  title: z.string(),
  body: z.string().nullable(),
  status: StatusSchema,
  position: z.number().nullable(),
  // The server emits `attributes` as an arbitrary JSON object (or null);
  // `z.record(z.string(), z.unknown())` is the zod 4 form (single-arg
  // `z.record` was removed — it now requires both key and value schemas).
  attributes: z.record(z.string(), z.unknown()).nullable(),
  relevance: RelevanceSchema.nullable(),
  effort: EffortSchema.nullable(),
  complexity: ComplexitySchema.nullable(),
  origin: OriginSchema.nullable(),
  closure_gate: ClosureGateSchema.nullable(),
  blocked_by_question_id: z.string().nullable(),
  enabling_option_id: z.string().nullable(),
  // Round-4 additions (T7). `.nullable()` is correct: domain.rs has zero
  // `skip_serializing_if` attributes (grep-verified during T7), so an unset
  // column is emitted as JSON `null` rather than as an absent key.
  task_kind: TaskKindSchema.nullable(),
  tier: TierSchema.nullable(),
  created_at: z.string(),
  updated_at: z.string(),
})

/**
 * A work item as returned by the tree endpoint: the same fields as
 * {@link WorkItem} plus its recursively-nested children.
 */
export interface WorkItemNode extends WorkItem {
  children: WorkItemNode[]
}

// Recursive schema. Zod 4 supports recursion via the JS getter pattern: the
// returned `WorkItemNodeSchema` is a plain `ZodObject` (full access to .pick /
// .extend / etc.) and `z.infer` resolves the self-reference correctly. This
// supersedes the zod-3 `z.lazy(() => ...)` form, which is still supported but
// no longer the canonical idiom.
export const WorkItemNodeSchema: z.ZodType<WorkItemNode> = WorkItemSchema.extend({
  get children() {
    return z.array(WorkItemNodeSchema)
  },
})

// ---------------------------------------------------------------------------
// Aggregate row types referenced by WorkItemDetail.
//
// T9/T10/T11 will MOVE each of these blocks to its own per-family file under
// `lumina/web/src/api/`; they live here for now so `WorkItemDetailWireSchema`
// has its dependencies in-scope without a forward-import chain.
// ---------------------------------------------------------------------------

export interface Finding {
  id: string
  work_item_id: string | null
  kind: string
  severity: Severity
  effort: string | null
  category: string
  status: string
  file: string | null
  line: number | null
  symbol: string | null
  summary: string
  description: string | null
  first_flagged: string | null
  rounds: number | null
  fingerprint: string | null
  flow: string | null
  dedup_id: string | null
  origin: Origin | null
  confidence: Confidence | null
  superseded_by: string | null
  resolved_at: string | null
  resolution: string | null
  defer_reason: string | null
  defer_trigger: string | null
  wontfix_rationale: string | null
}

export const FindingSchema = z.object({
  id: z.string(),
  work_item_id: z.string().nullable(),
  kind: z.string(),
  severity: SeveritySchema,
  effort: z.string().nullable(),
  category: z.string(),
  status: z.string(),
  file: z.string().nullable(),
  line: z.number().nullable(),
  symbol: z.string().nullable(),
  summary: z.string(),
  description: z.string().nullable(),
  first_flagged: z.string().nullable(),
  rounds: z.number().nullable(),
  fingerprint: z.string().nullable(),
  flow: z.string().nullable(),
  dedup_id: z.string().nullable(),
  origin: OriginSchema.nullable(),
  confidence: ConfidenceSchema.nullable(),
  superseded_by: z.string().nullable(),
  resolved_at: z.string().nullable(),
  // `resolution` is free-text describing the change that fixed the finding
  // (e.g. "Added exp check in refresh handler"), NOT a Disposition value. The
  // disposition vocabulary belongs on `status` above. Keep this as a plain
  // nullable string to match the wire.
  resolution: z.string().nullable(),
  defer_reason: z.string().nullable(),
  defer_trigger: z.string().nullable(),
  wontfix_rationale: z.string().nullable(),
  // Migration 0004: `findings.repo_id` — nullable foreign key into `repo_links`.
  // `nullable()` handles the live wire shape (column is nullable, NULL = the
  // file lives in the project's primary repo); `optional()` tolerates absent
  // fields from any pre-deploy caches that wouldn't yet emit the key.
  repo_id: z.string().nullable().optional(),
})

export interface ContextBlock {
  id: string
  title: string
  body: string
  created_at: string
  updated_at: string
}

export const ContextBlockSchema = z.object({
  id: z.string(),
  title: z.string(),
  body: z.string(),
  created_at: z.string(),
  updated_at: z.string(),
})

/**
 * A row of `acceptance_criteria` (migration 0003). Note: on the wire `checked`
 * is a 0/1 integer (the Rust side mirrors the SQLite INTEGER column as `i64`),
 * but this boundary normalises it to a JS `boolean` in {@link fetchDetail} so
 * consumers can use truthy semantics directly.
 */
export interface AcceptanceCriterion {
  id: string
  work_item_id: string
  seq: number
  text: string
  checked: boolean
  checked_at: string | null
  checked_by: string | null
  created_at: string
}

// Wire shape: `checked` arrives as 0/1 integer; the normalised boolean form
// (exposed as {@link AcceptanceCriterion}) is produced by fetchDetail's
// post-parse transform.
export const AcceptanceCriterionWireSchema = z.object({
  id: z.string(),
  work_item_id: z.string(),
  seq: z.number(),
  text: z.string(),
  checked: z.number(),
  checked_at: z.string().nullable(),
  checked_by: z.string().nullable(),
  created_at: z.string(),
})

/** A row of `research_notes` (migration 0003). */
export interface ResearchNote {
  id: string
  work_item_id: string
  seq: number
  summary: string
  body: string | null
  confidence: Confidence | null
  state: ResearchState | null
  rationale: string | null
  lens: string | null
  origin: Origin | null
  superseded_by: string | null
  created_at: string
}

export const ResearchNoteSchema = z.object({
  id: z.string(),
  work_item_id: z.string(),
  seq: z.number(),
  summary: z.string(),
  body: z.string().nullable(),
  confidence: ConfidenceSchema.nullable(),
  state: ResearchStateSchema.nullable(),
  rationale: z.string().nullable(),
  lens: z.string().nullable(),
  origin: OriginSchema.nullable(),
  superseded_by: z.string().nullable(),
  created_at: z.string(),
})

/** A row of `question_options` (migration 0003): one branch of an open question. */
export interface QuestionOption {
  id: string
  question_id: string
  seq: number
  label: string
  detail: string | null
  created_at: string
}

export const QuestionOptionSchema = z.object({
  id: z.string(),
  question_id: z.string(),
  seq: z.number(),
  label: z.string(),
  detail: z.string().nullable(),
  created_at: z.string(),
})

/** A row of `open_questions` (migration 0003): a story-scoped decision. */
export interface OpenQuestion {
  id: string
  story_id: string
  seq: number
  question: string
  status: QuestionStatus | null
  answer: string | null
  chosen_option_id: string | null
  decided_at: string | null
  decided_by: string | null
  prompting_finding_id: string | null
  prompting_note_id: string | null
  created_at: string
  options: QuestionOption[]
}

export const OpenQuestionSchema = z.object({
  id: z.string(),
  story_id: z.string(),
  seq: z.number(),
  question: z.string(),
  status: QuestionStatusSchema.nullable(),
  answer: z.string().nullable(),
  chosen_option_id: z.string().nullable(),
  decided_at: z.string().nullable(),
  decided_by: z.string().nullable(),
  prompting_finding_id: z.string().nullable(),
  prompting_note_id: z.string().nullable(),
  created_at: z.string(),
  options: z.array(QuestionOptionSchema),
})

/** A row of `work_item_activity` (migration 0002): the per-item activity log. */
export interface WorkItemActivity {
  id: string
  work_item_id: string
  seq: number
  entry_kind: ActivityType
  author: string | null
  summary: string
  payload: Record<string, unknown> | null
  origin: Origin | null
  created_at: string
}

export const WorkItemActivitySchema = z.object({
  id: z.string(),
  work_item_id: z.string(),
  seq: z.number(),
  entry_kind: ActivityTypeSchema,
  author: z.string().nullable(),
  summary: z.string(),
  payload: z.record(z.string(), z.unknown()).nullable(),
  origin: OriginSchema.nullable(),
  created_at: z.string(),
})

// ---------------------------------------------------------------------------
// Round-4 additions (T7): risks, rejected alternatives, task dependencies.
//
// Mirrors `domain::Risk` / `domain::RejectedAlternative` / `domain::TaskDependency`.
// T10 will move these to per-family files (`risks.ts`, `rejected-alternatives.ts`,
// `task-deps.ts`) and re-import from here.
// ---------------------------------------------------------------------------

/** A row of `risks` (migration 0005): a per-work-item risk register entry. */
export interface Risk {
  id: string
  work_item_id: string
  seq: number
  summary: string
  body: string | null
  rationale: string | null
  /**
   * CHECK-enforced `low|medium|high|critical`. Carried as `Option<String>` on
   * the Rust row to match the codebase's "row stores plain string, MCP-param
   * layer carries the typed enum" idiom; we mirror that here as a
   * `RiskSeverity | null`. In practice NOT NULL on the column, so it should
   * always be set on read — but we accept null to mirror the Rust shape.
   */
  severity: RiskSeverity | null
  mitigation: string | null
  superseded_by: string | null
  created_at: string
}

export const RiskSchema = z.object({
  id: z.string(),
  work_item_id: z.string(),
  seq: z.number(),
  summary: z.string(),
  body: z.string().nullable(),
  rationale: z.string().nullable(),
  severity: RiskSeveritySchema.nullable(),
  mitigation: z.string().nullable(),
  superseded_by: z.string().nullable(),
  created_at: z.string(),
})

/** A row of `rejected_alternatives` (migration 0005): a discarded planning option. */
export interface RejectedAlternative {
  id: string
  work_item_id: string
  seq: number
  summary: string
  body: string | null
  rationale: string | null
  /**
   * Free-text confidence grade (`high|medium|low` validated in the repo,
   * mirroring `research_notes.confidence`). Free TEXT on the Rust side, so we
   * type it as `Confidence | null` to surface the closed enum at the wire
   * while still tolerating any historical free-text values.
   */
  confidence: Confidence | null
  superseded_by: string | null
  created_at: string
}

export const RejectedAlternativeSchema = z.object({
  id: z.string(),
  work_item_id: z.string(),
  seq: z.number(),
  summary: z.string(),
  body: z.string().nullable(),
  rationale: z.string().nullable(),
  confidence: ConfidenceSchema.nullable(),
  superseded_by: z.string().nullable(),
  created_at: z.string(),
})

/** A row of `task_dependencies` (migration 0005): a task→task prerequisite edge. */
export interface TaskDependency {
  task_id: string
  depends_on_id: string
  /** Edge category — `data|sequence|…`; free TEXT, default `'data'`. */
  kind: string
  created_at: string
}

export const TaskDependencySchema = z.object({
  task_id: z.string(),
  depends_on_id: z.string(),
  kind: z.string(),
  created_at: z.string(),
})

// ---------------------------------------------------------------------------
// WorkItemDetail (the GET /work-items/{id} response).
// ---------------------------------------------------------------------------

/** Response shape of `GET /api/work-items/{id}` (WorkItemDetail). */
export interface WorkItemDetail {
  item: WorkItem
  children: WorkItem[]
  findings: Finding[]
  context_blocks: ContextBlock[]
  activity: WorkItemActivity[]
  acceptance_criteria: AcceptanceCriterion[]
  research_notes: ResearchNote[]
  open_questions: OpenQuestion[]
  // Migration 0004: populated only when `item.kind === 'project'`; defaults to
  // [] for every other kind. Rust side uses `#[serde(default)]`, so the field
  // may also be absent from any pre-deploy cached responses.
  repo_links: RepoLink[]
  // Migration 0005 (round-4 T7): risks, rejected alternatives, task
  // dependencies. Same `#[serde(default)]` semantics → `.default([])`.
  risks: Risk[]
  rejected_alternatives: RejectedAlternative[]
  task_dependencies: TaskDependency[]
}

// Wire-shape schema for the detail endpoint: identical to the consumer-facing
// {@link WorkItemDetail} except `acceptance_criteria[].checked` is an integer
// (0/1). fetchDetail parses this shape, then runs a single-field transform to
// produce the boolean-normalised consumer type.
export const WorkItemDetailWireSchema = z.object({
  item: WorkItemSchema,
  children: z.array(WorkItemSchema),
  findings: z.array(FindingSchema),
  context_blocks: z.array(ContextBlockSchema),
  activity: z.array(WorkItemActivitySchema),
  acceptance_criteria: z.array(AcceptanceCriterionWireSchema),
  research_notes: z.array(ResearchNoteSchema),
  open_questions: z.array(OpenQuestionSchema),
  // `optional().default([])` matches the Rust `#[serde(default)]` semantics so
  // the schema accepts both responses that omit the field (non-project items
  // and pre-deploy cached responses) and explicit empty arrays.
  repo_links: z.array(RepoLinkSchema).optional().default([]),
  // Round-4 (T7): the three new aggregates. Same `optional().default([])`
  // contract as `repo_links` — backend stamps `#[serde(default)]` on each.
  risks: z.array(RiskSchema).optional().default([]),
  rejected_alternatives: z.array(RejectedAlternativeSchema).optional().default([]),
  task_dependencies: z.array(TaskDependencySchema).optional().default([]),
})

// ---------------------------------------------------------------------------
// Fetch wrappers.
// ---------------------------------------------------------------------------

/** Body accepted by `POST /api/work-items`. */
export interface CreateWorkItemRequest {
  kind: Kind
  parent_id?: string | null
  title: string
  body?: string | null
}

/**
 * `GET /api/work-items` — the full nested tree of ROOT nodes (each with a
 * recursive `children` array).
 */
export async function fetchTree(): Promise<WorkItemNode[]> {
  return handle<WorkItemNode[]>(
    await fetch(`${API_BASE}/work-items`),
    z.array(WorkItemNodeSchema),
  )
}

/** `GET /api/work-items/{id}` — item + children + findings + context blocks.
 *
 * Normalises the wire-level 0/1 integer `acceptance_criteria[].checked` into a
 * JS boolean at this boundary so downstream consumers can use truthy semantics
 * directly (rather than `=== 1`). All other fields pass through unchanged.
 */
export async function fetchDetail(id: string): Promise<WorkItemDetail> {
  const wire = await handle(
    await fetch(`${API_BASE}/work-items/${encodeURIComponent(id)}`),
    WorkItemDetailWireSchema,
  )
  return {
    ...wire,
    acceptance_criteria: wire.acceptance_criteria.map((ac) => ({
      ...ac,
      // Wire is 0/1 (SQLite INTEGER mirrored as Rust i64); strict `=== 1`
      // rather than truthy so a stray non-1 numeric does not render as ticked.
      checked: ac.checked === 1,
    })),
  }
}

/** `POST /api/work-items` — create a node; returns the created work item. */
export async function createWorkItem(req: CreateWorkItemRequest): Promise<WorkItem> {
  return handle<WorkItem>(
    await fetch(`${API_BASE}/work-items`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(req),
    }),
    WorkItemSchema,
  )
}

/**
 * Body accepted by `PATCH /api/work-items/{id}` (mirrors `domain::UpdateWorkItemRequest`).
 * Every field is optional with SET-OR-LEAVE semantics: an absent field leaves
 * the column untouched (the repo's `COALESCE(?, col)` write), it does NOT clear
 * the column to NULL.
 */
export interface UpdateWorkItemRequest {
  title?: string
  body?: string
  status?: Status
  position?: number
  attributes?: Record<string, unknown>
}

/**
 * `PATCH /api/work-items/{id}` — general partial-update; any subset of
 * `title`/`body`/`status`/`position`/`attributes` may be supplied.
 */
export async function updateWorkItem(
  id: string,
  patch: UpdateWorkItemRequest,
): Promise<WorkItem> {
  return handle<WorkItem>(
    await fetch(`${API_BASE}/work-items/${encodeURIComponent(id)}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(patch),
    }),
    WorkItemSchema,
  )
}

/**
 * `PATCH /api/work-items/{id}` — narrow status-only wrapper. Delegates to
 * {@link updateWorkItem} so both share the same fetch + validation path; kept
 * as a named export because existing call sites pass a free-text status string.
 */
export async function updateStatus(id: string, status: string): Promise<WorkItem> {
  return updateWorkItem(id, { status: status as Status })
}

File: lumina/src/domain/work_items.rs
//! Work-item row + detail aggregate types, the PTY-session row structs, context
//! blocks, and the work-item create/update request bodies. Carved out of
//! `domain/mod.rs` (D1 refactor); re-exported via `pub use work_items::*`.

use super::*;
use serde::{Deserialize, Serialize};

/// A row of `work_items`. The 5-level hierarchy (`project > epic > focus >
/// story > task`) is an adjacency list via `parent_id`.
#[derive(Debug, Clone, Serialize)]
pub struct WorkItem {
    pub id: String,
    pub kind: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parent_id: Option<String>,
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub body: Option<String>,
    pub status: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub position: Option<i64>,
    /// Nullable JSON object of kind-specific fields (migration 0002); `None`
    /// means "no kind-specific fields".
    #[serde(skip_serializing_if = "Option::is_none")]
    pub attributes: Option<serde_json::Value>,
    /// Relevance axis (migration 0003): `active|backlog|deferred|rejected`. Set
    /// only on epic/focus/story; NULL on task/project.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub relevance: Option<String>,
    /// Effort grade (migration 0003): `s|m|l` (task scope); NULL otherwise.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub effort: Option<String>,
    /// Complexity grade (migration 0003): `low|medium|high` (task scope).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub complexity: Option<String>,
    /// Provenance (migration 0003): which command produced this item.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub origin: Option<String>,
    /// Per-story closure gate (migration 0003): `hard|soft` (story scope).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub closure_gate: Option<String>,
    /// Task is blocked while this `open_questions` row is open (migration 0003).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub blocked_by_question_id: Option<String>,
    /// Task is exclusive to this `question_options` branch (migration 0003).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub enabling_option_id: Option<String>,
    /// Task-scope phase-disposition (migration 0005 + 0007 cull):
    /// `foundation|main|polish`. NULL on non-task rows; the repo layer is the
    /// source of truth for the "task rows only" rule (no DB-level kind coupling).
    /// Mirrors the `effort`/`complexity` idiom — stored as `Option<String>` on
    /// the row, with the typed [`TaskKind`] enum used by the wire / MCP layer.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub task_kind: Option<String>,
    /// Dispatch tier (migration 0006): `lite|deep`. NULL on non-task rows; the
    /// repo layer is the source of truth for the "task rows only" rule (no
    /// DB-level kind coupling). Mirrors the `task_kind`/`effort`/`complexity`
    /// idiom — stored as `Option<String>` on the row, with the typed [`Tier`]
    /// enum used by the wire / MCP layer.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tier: Option<String>,
    /// Focus shape (migration 0010): `vertical-slice|cross-cutting|foundational`.
    /// Set only on `focus` rows, NULL otherwise; the repo layer is the source of
    /// truth for the focus-only rule (no DB-level kind coupling). Mirrors the
    /// `task_kind`/`tier` idiom — stored as `Option<String>` on the row, with the
    /// typed [`Shape`] enum used by the wire / MCP layer.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub shape: Option<String>,
    /// Provenance back-link (migration 0011): the `findings.id` a triage decision
    /// spawned this work item from; NULL on items not spawned from a finding. A
    /// scalar (placed before the timestamp scalars, NOT after any Vec field) so
    /// the export tables-last ordering gate stays satisfied.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub spawned_from_finding_id: Option<String>,
    /// Work-queue claim owner (team-execution migration): the agent id holding
    /// the current lease on this task; NULL when unclaimed. A scalar placed
    /// before the timestamp scalars (NOT after any Vec field) so the export
    /// tables-last ordering gate stays satisfied.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub assignee: Option<String>,
    /// Work-queue lease deadline (team-execution migration): ISO-8601 instant
    /// after which the lease is reclaimable; NULL when unclaimed.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub lease_expires_at: Option<String>,
    /// Work-queue lane (team-execution migration): `implement|review`. NULL on
    /// rows outside the queue. Carried as `Option<String>` on the row per the
    /// row-struct idiom (see `tier`/`task_kind`), with the typed [`Lane`] enum
    /// used by the wire / MCP / claim-result layer.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub lane: Option<String>,
    /// Review back-link (team-execution migration): for a `lane='review'` task,
    /// the `work_items.id` of the implementation task this review covers; NULL
    /// otherwise.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reviews_work_item_id: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    /// Soft-delete tombstone instant (`None` = live). Carried here off the detail
    /// / list row read so the export tombstone fold reads it from the already
    /// fetched detail rather than issuing a separate `deleted_at` query (O17).
    /// `#[serde(skip_serializing)]` keeps it OFF both the JSON wire and the TOML
    /// export (the drain re-inserts a top-level `deleted_at` tombstone key itself),
    /// so adding this field changes no public contract.
    #[serde(skip_serializing)]
    pub deleted_at: Option<String>,
}

/// A row of `work_item_activity` (migration 0002): the append-only per-item
/// activity log, ordered by the per-item monotonic `seq`. Read aggregate only —
/// `Serialize` but not `JsonSchema` (mirrors `WorkItem`/`Finding`).
#[derive(Debug, Clone, Serialize)]
pub struct WorkItemActivity {
    pub id: String,
    pub work_item_id: String,
    pub seq: i64,
    pub entry_kind: String,
    pub author: Option<String>,
    pub summary: String,
    pub payload: Option<serde_json::Value>,
    /// Provenance (migration 0003): which command produced this activity entry.
    pub origin: Option<String>,
    pub created_at: String,
}

/// A row of `pty_sessions` (migration 0008): a supervised `claude` PTY
/// session, carrying lifecycle status, the spawn-config snapshot, and the
/// parse-strategy version pin. Read aggregate only — `Serialize`, no
/// `JsonSchema` (mirrors the other row structs; create/update bodies live
/// as separate types when needed).
///
/// `status` is one of `spawning|active|idle|awaiting|completed|failed|cancelled`
/// (free TEXT on the column — the typed enum is `pty::protocol::SessionStatus`
/// at the wire layer). `parse_strategy_version` is the supervisor's parser
/// generation (defaults to 1 on insert; bumped when a future parse-strategy
/// change requires a session reset).
#[derive(Debug, Clone, Serialize)]
pub struct PtySession {
    pub id: String,
    pub label: Option<String>,
    pub project_id: Option<String>,
    pub cwd: String,
    pub config_json: String,
    pub parse_strategy_version: i64,
    pub status: String,
    pub started_at: String,
    pub updated_at: String,
    pub ended_at: Option<String>,
    pub exit_code: Option<i64>,
    pub last_error: Option<String>,
    pub previous_session_id: Option<String>,
    pub jsonl_path: Option<String>,
}

/// A row of `pty_messages` (migration 0008): one ordered transcript entry on
/// a PTY session, with `sequence` monotone within the session and `kind`
/// drawn from `user_input|assistant_text|tool_use|tool_result|system|error`
/// (free TEXT on the column — typed at the wire layer via
/// `pty::protocol::MessageKind`). The pre-jsonl-tail vocab carried
/// `tool_call|prompt|parser_unknown` as well; those three were retired in T5
/// of `lumina-pty-jsonl-tail` when the vt100 row-finalisation parser was
/// replaced by JSONL-tail message extraction. `content_json` carries the
/// per-kind payload; `raw_text` is the ansi-stripped fallback retained for
/// replay / debugging.
#[derive(Debug, Clone, Serialize)]
pub struct PtyMessage {
    pub id: String,
    pub session_id: String,
    pub sequence: i64,
    pub created_at: String,
    pub kind: String,
    pub content_json: String,
    pub raw_text: Option<String>,
}

/// A row of `pty_queue` (migration 0008): one pending or in-flight client
/// input frame waiting to be dispatched to the supervisor. `sequence` is
/// monotone within the session; `input_kind` is `prompt|cancel|control` (free
/// TEXT; typed via `pty::protocol::InputKind`); `status` walks
/// `pending → dispatched → completed|failed|cancelled`.
#[derive(Debug, Clone, Serialize)]
pub struct PtyQueueEntry {
    pub id: String,
    pub session_id: String,
    pub sequence: i64,
    pub input_kind: String,
    pub payload: String,
    pub enqueued_at: String,
    pub dispatched_at: Option<String>,
    pub completed_at: Option<String>,
    pub status: String,
    pub error: Option<String>,
}

/// A row of `context_blocks` — the drift-killer. Shared context is one row
/// referenced by many work-items through `work_item_context`.
#[derive(Debug, Clone, Serialize)]
pub struct ContextBlock {
    pub id: String,
    pub title: Option<String>,
    pub body: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

/// Read-aggregate for the detail endpoint: an item plus its DIRECT children,
/// its findings, and its linked context blocks. The full tree is assembled by
/// the HTTP layer / frontend from repeated `list_work_items` calls — direct
/// children are sufficient for the slice.
#[derive(Debug, Clone, Serialize)]
pub struct WorkItemDetail {
    pub item: WorkItem,
    pub children: Vec<WorkItem>,
    pub findings: Vec<Finding>,
    pub context_blocks: Vec<ContextBlock>,
    /// The item's activity-log rows (migration 0002), ordered by `seq`.
    pub activity: Vec<WorkItemActivity>,
    /// The item's acceptance criteria (migration 0003), ordered by `seq`.
    pub acceptance_criteria: Vec<AcceptanceCriterion>,
    /// The item's research notes (migration 0003) — Task 4 implements the fold.
    pub research_notes: Vec<ResearchNote>,
    /// The item's open questions (migration 0003) — Task 4 implements the fold.
    pub open_questions: Vec<OpenQuestion>,
    /// Linked GitHub repos (migration 0004) — populated only when
    /// `item.kind == "project"`; empty otherwise.
    #[serde(default)]
    pub repo_links: Vec<RepoLink>,
    /// The item's risk register (migration 0005), ordered by `seq`. Repo layer
    /// implements the fold (Task 3).
    #[serde(default)]
    pub risks: Vec<Risk>,
    /// The item's rejected planning alternatives (migration 0005), ordered by
    /// `seq`. Repo layer implements the fold (Task 3).
    #[serde(default)]
    pub rejected_alternatives: Vec<RejectedAlternative>,
    /// Outgoing task→task dependency edges (migration 0005) — populated only
    /// when `item.kind == "task"`; empty otherwise. Repo layer (Task 3) is the
    /// source of truth for the kind filter; an empty vec is the
    /// not-applicable state for non-task rows.
    #[serde(default)]
    pub task_dependencies: Vec<TaskDependency>,
}

/// Create-body for a new work item. Deserialised by the HTTP POST handler
/// (Task 4) and the MCP `create_work_item` tool (Task 5). `JsonSchema` is
/// derived for the rmcp `Parameters<T>` tool-argument contract.
#[derive(Debug, Clone, Deserialize, schemars::JsonSchema)]
pub struct CreateWorkItemRequest {
    /// One of `project`/`epic`/`focus`/`story`/`task`.
    pub kind: String,
    /// Parent work-item id; `None`/absent only for a `project`.
    #[serde(default)]
    pub parent_id: Option<String>,
    pub title: String,
    #[serde(default)]
    pub body: Option<String>,
    /// Provenance (migration 0003): which command produced this item
    /// (`plan|implement|review|optimise|tdd|human|none`); absent ⇒ NULL.
    #[serde(default)]
    pub origin: Option<String>,
    /// Epic outcome statement (migration 0010): MANDATORY for an `epic` at
    /// create (a non-empty value), rejected on any non-epic kind. Folded into
    /// the new epic's `attributes` JSON by `repo::create_work_item_full`.
    #[serde(default)]
    pub outcome: Option<String>,
    /// Focus shape (migration 0010): MANDATORY for a `focus` at create, rejected
    /// on any non-focus kind. Bound to `work_items.shape` by
    /// `repo::create_work_item_full`.
    #[serde(default)]
    pub shape: Option<String>,
}

/// Revise-later plan body for an epic (migration 0010). Both fields are
/// optional with present-only JSON-merge semantics (an absent field leaves the
/// stored attribute untouched). Consumed by the HTTP PATCH handler (T7) and the
/// MCP `set_epic_plan` tool (T6), both delegating to `repo::set_epic_plan`.
#[derive(Debug, Clone, Deserialize, schemars::JsonSchema)]
pub struct EpicPlanRequest {
    /// Epic outcome statement; absent leaves the stored value untouched.
    #[serde(default)]
    pub outcome: Option<String>,
    /// Epic context note; absent leaves the stored value untouched.
    #[serde(default)]
    pub context: Option<String>,
}

/// Revise-later plan body for a focus (migration 0010). The single field is
/// optional with present-only JSON-merge semantics. Consumed by the HTTP PATCH
/// handler (T7) and the MCP `set_focus_plan` tool (T6), both delegating to
/// `repo::set_focus_plan`.
#[derive(Debug, Clone, Deserialize, schemars::JsonSchema)]
pub struct FocusPlanRequest {
    /// Focus framing statement; absent leaves the stored value untouched.
    #[serde(default)]
    pub framing: Option<String>,
}

/// Update-body for a status transition. Deserialised by the HTTP PATCH handler
/// (Task 4) and the MCP `update_work_item_status` tool (Task 5).
#[derive(Debug, Clone, Deserialize, schemars::JsonSchema)]
pub struct UpdateStatusRequest {
    pub status: String,
}

/// Partial-update body for a work item. Every field is optional with
/// SET-OR-LEAVE semantics: an absent/`None` field leaves the column untouched
/// (the repo's `COALESCE(?, col)` write), it does NOT clear the column to NULL.
/// Deserialised by the HTTP PATCH handler (Task 4) and the MCP update tool
/// (Task 5); `JsonSchema` is derived for the rmcp `Parameters<T>` contract.
#[derive(Debug, Clone, Deserialize, schemars::JsonSchema)]
pub struct UpdateWorkItemRequest {
    /// New title; absent leaves the existing title unchanged.
    #[serde(default)]
    pub title: Option<String>,
    /// New body; absent leaves the existing body unchanged (does NOT clear it).
    #[serde(default)]
    pub body: Option<String>,
    /// New status; absent leaves the existing status unchanged.
    #[serde(default)]
    pub status: Option<Status>,
    /// New sibling-ordering position; absent leaves the existing position unchanged.
    #[serde(default)]
    pub position: Option<i64>,
    /// New kind-specific attributes JSON object; absent leaves the existing
    /// attributes unchanged (does NOT clear them).
    #[serde(default)]
    pub attributes: Option<serde_json::Value>,
}
